# **Exploratory Data Analisys: Sectors**

## **Initial Setup**

### Install Packages

In [25]:
%pip install pandas -q
%pip install plotly -q

Note: you may need to restart the kernel to use updated packages.
Note: you may need to restart the kernel to use updated packages.


### Import libs

In [26]:
import os
import itertools
import pandas as pd
from pathlib import Path
import plotly.express as px
import plotly.graph_objects as go
import plotly.subplots as sp

### Pandas Config

In [27]:
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

### Create a file path default

In [28]:
file_path_book = str(Path(os.getcwd()).parent.parent.parent / 'data/book')

## **Fundamentals**

#### Load data

In [29]:
df_fundamentals_book = pd.read_csv(file_path_book + '/fundamentals_book.csv')
df_fundamentals_book.head(2)

ticker                                          long_name  \
0  ABCB4.SA                              Banco ABC Brasil S.A.   
1  AGRO3.SA  BrasilAgro - Companhia Brasileira de Proprieda...   

               sector          industry    market_cap  enterprise_value  \
0  Financial Services  Banks - Regional  4.265434e+09      1.477339e+10   
1  Consumer Defensive     Farm Products  2.466480e+09      2.912933e+09   

   total_revenue  profit_margins  operating_margins  dividend_rate   beta  \
0   1.941779e+09         0.41576            0.38826           1.56  0.679   
1   1.249437e+09         0.21493            0.25031           3.21  0.432   

        ebitda  trailing_pe  forward_pe    volume  average_volume  \
0          0.0     4.069768    4.706601   92300.0        747165.0   
1  264892000.0     9.450382    6.332481  298100.0        666692.0   

   fifty_two_week_low  fifty_two_week_high  price_to_sales_trailing_12_months  \
0               15.85                21.99                           2.196663   
1               22.29                32.71                           1.974073   

   fifty_day_average  two_hundred_day_average  trailing_annual_dividend_rate  \
0            19.3382                 18.14667                           1.55   
1            27.0106                 25.58635                           3.24   

   trailing_annual_dividend_yield  book_value  price_to_book    total_cash  \
0                        0.080687      24.518       0.785138  7.774306e+09   
1                        0.132029      22.237       1.113460  3.838370e+08   

   total_cash_per_share    total_debt  earnings_quarterly_growth  \
0                35.162  1.829846e+10                      0.001   
1                 3.885  8.720750e+08                      6.801   

   revenue_growth  gross_margins  ebitda_margins  return_on_assets  \
0           0.003        0.00000         0.00000           0.01530   
1           0.671        0.25252         0.21201           0.03839   

   return_on_equity  gross_profits  total_assets_approx  asset_turnover  \
0            0.1568   1.973086e+09         7.774306e+09        0.249769   
1            0.1217   3.155040e+08         3.838370e+08        3.255124   

   earnings_growth_rate  dividend_payout_ratio        equity  debt_to_equity  \
0                   0.1          155000.000000 -1.052416e+10       -1.738710   
1                 680.1              47.640053 -4.882380e+08       -1.786168   

        roi      roce  
0  0.131438  0.000000  
1  0.428927  0.079343

In [30]:
df_fundamentals_numeric_cols = df_fundamentals_book.select_dtypes(include=['int', 'number', 'float64'])
df_fundamentals_numeric_cols.head(2)

market_cap  enterprise_value  total_revenue  profit_margins  \
0  4.265434e+09      1.477339e+10   1.941779e+09         0.41576   
1  2.466480e+09      2.912933e+09   1.249437e+09         0.21493   

   operating_margins  dividend_rate   beta       ebitda  trailing_pe  \
0            0.38826           1.56  0.679          0.0     4.069768   
1            0.25031           3.21  0.432  264892000.0     9.450382   

   forward_pe    volume  average_volume  fifty_two_week_low  \
0    4.706601   92300.0        747165.0               15.85   
1    6.332481  298100.0        666692.0               22.29   

   fifty_two_week_high  price_to_sales_trailing_12_months  fifty_day_average  \
0                21.99                           2.196663            19.3382   
1                32.71                           1.974073            27.0106   

   two_hundred_day_average  trailing_annual_dividend_rate  \
0                 18.14667                           1.55   
1                 25.58635                           3.24   

   trailing_annual_dividend_yield  book_value  price_to_book    total_cash  \
0                        0.080687      24.518       0.785138  7.774306e+09   
1                        0.132029      22.237       1.113460  3.838370e+08   

   total_cash_per_share    total_debt  earnings_quarterly_growth  \
0                35.162  1.829846e+10                      0.001   
1                 3.885  8.720750e+08                      6.801   

   revenue_growth  gross_margins  ebitda_margins  return_on_assets  \
0           0.003        0.00000         0.00000           0.01530   
1           0.671        0.25252         0.21201           0.03839   

   return_on_equity  gross_profits  total_assets_approx  asset_turnover  \
0            0.1568   1.973086e+09         7.774306e+09        0.249769   
1            0.1217   3.155040e+08         3.838370e+08        3.255124   

   earnings_growth_rate  dividend_payout_ratio        equity  debt_to_equity  \
0                   0.1          155000.000000 -1.052416e+10       -1.738710   
1                 680.1              47.640053 -4.882380e+08       -1.786168   

        roi      roce  
0  0.131438  0.000000  
1  0.428927  0.079343

### Sector Analisys

#### Number of companies by sector

In [31]:
sector_counts = df_fundamentals_book['sector'].value_counts(ascending=True)

fig = px.bar(x=sector_counts.index, y=sector_counts.values, title='Number of companies by sector',color_discrete_sequence=['rgb(100, 195, 181)']*len(sector_counts), template='plotly_dark', hover_name=sector_counts.index, height=485)
fig.update_traces(text=sector_counts.values.astype(str), textposition='outside',textfont=dict(color='white'))
fig.update_layout(xaxis_title='Sector', yaxis_title='Number of companies')
fig.show()

- **Companies Listed by Sector:**
  - `Technology`: 4 companies, the sector with the least representation, potential for growth.
  - `Communication Services`: 7 companies, including telecommunications and media.
  - `Energy`: 9 companies, essential and stable, encompasses oil, gas, and renewable energy.
  - `Healthcare`: 13 companies, growth reflects innovation and demographics.
  - `Consumer Defensive and Real Estate`: 21 companies each, stability during times of uncertainty.
  - `Basic Materials`: 29 companies, crucial for construction and manufacturing.
  - `Financial Services`: 40 companies, diverse and vital to the economy.
  - `Utilities`: 46 companies, reflect demand for basic services.
  - `Industrials`: 48 companies, significant diversity and expansion.
  - `Consumer Cyclical`: 53 companies, the largest sector, indicative of consumer confidence.
  
- **Observations:**
  - The `Consumer Cyclical` sector has the highest representation, suggesting an economy geared towards consumer spending and a trend towards private consumption growth.
  - The `Technology sector`, despite being the least represented, plays a critical and increasingly influential role in the economy, pointing to an area for future observation and investment.
  - The `Industrials` and `Services` sectors show strong presence, reflecting the importance of consistent operations and infrastructure in the economy.
  - `Consumer Defensive` and `Real Estate` have equal representation, illustrating their perception as safe havens for investors.

These data not only indicate the current composition of the sectors on the stock exchange but also point towards potential areas of growth and future economic development.


#### Outlier Analysis

In [32]:
num_columns = len(df_fundamentals_numeric_cols)
num_rows = num_columns // 3 + (num_columns % 3 > 0)

sector_means = df_fundamentals_numeric_cols.groupby(df_fundamentals_book['sector']).mean()

subplot_titles = [str(col) for col in sector_means]

fig = sp.make_subplots(rows=num_rows, cols=3, subplot_titles=subplot_titles)

for i, column in enumerate(sector_means, start=1):

    row = (i - 1) // 3 + 1
    col = (i - 1) % 3 + 1
    
    trace = go.Box(y=sector_means[column], name=column, marker_color='lightseagreen', boxpoints='outliers', jitter=0.7, hoverinfo='y+text',text=(df_fundamentals_book['sector'] + ' - ' + df_fundamentals_book['long_name']),)

    fig.add_trace(trace, row=row, col=col)

fig.update_layout(title_text='Boxplot of Numerical Variables by Sector', height=300*num_rows, showlegend=False, template='plotly_dark')
fig.show()


##### Detailed View of Outliers by Column

**Basic Materials Sector:**

 - `earnings_quarterly_growth`: **67.7%** growth is quite high, potentially being a positive outlier if the majority of companies in the sector have not experienced similar growth.

 - r`evenue_growth`: A revenue reduction of **-22.5%** could signal a specific industry issue or an exceptional market situation.

**Communication Services Sector:**

 - `profit_margins`: A profit margin of **-56.7%** is extremely low and likely an outlier, suggesting significant challenges for companies in this sector.

**Consumer Cyclical Sector:**

 - `operating_margins`: Operating margins of **-59.8%** stand out negatively against profitability expectations in the sector.

**Healthcare Sector:**

- `profit_margins` and `revenue_growth`: Previous data suggest extreme values that could be outliers when compared to the sector average.

**Technology Sector:**

 - `earnings_quarterly_growth`: As technology is a rapidly growing sector, a **67.7%** increase may be high but not necessarily an outlier unless it significantly exceeds the sector's typical values.

**Real Estate Sector:**

 - `operating_margins` and `revenue_growth`: Data indicate potential for outliers, especially if operating margins or revenue growth are atypical for the sector.

**Non-Cyclical Consumer Goods Sector:**

 - `dividend_yield` and `revenue_growth`: High dividend yields or negative revenue growth could be outliers depending on how they compare to sector standards.

#### Histogram and Dispersion

In [33]:
subplot_titles = [str(col) for col in df_fundamentals_numeric_cols]
columns_per_row = 3
num_rows = len(df_fundamentals_numeric_cols) // columns_per_row + (len(df_fundamentals_numeric_cols) % columns_per_row > 0)

fig = sp.make_subplots(rows=num_rows, cols=columns_per_row, subplot_titles=subplot_titles)

for i, column in enumerate(df_fundamentals_numeric_cols):

    row = i // columns_per_row + 1
    col = i % columns_per_row + 1
    
    fig.add_trace(go.Histogram(x=df_fundamentals_book[column],name=column, marker_color='lightseagreen'),row=row,col=col)

fig.update_layout(title = 'Histograms by Sector', height=300 * num_rows, showlegend=False, template='plotly_dark')
fig.show()

In [34]:
num_cols = 2
combinations = list(itertools.combinations(df_fundamentals_numeric_cols.columns, 2))
num_rows = (len(combinations) + num_cols - 1) // num_cols

df = df_fundamentals_numeric_cols.groupby(df_fundamentals_book['sector']).mean().reset_index()

fig = sp.make_subplots(rows=num_rows, cols=num_cols, subplot_titles=[f'{col1} vs {col2}' for col1, col2 in combinations])

for i, (col1, col2) in enumerate(combinations):
    row = i // num_cols + 1
    col = i % num_cols + 1

    scatter_fig = px.scatter(df, x=col1, y=col2, template='plotly_dark')
    scatter_traces = scatter_fig['data']

    for trace in scatter_traces:
        fig.add_trace(trace, row=row, col=col)

fig.update_layout(title='Scatter Plot Matrix', height=200 * num_rows, showlegend=False, template='plotly_dark')
fig.show()

##### Sector Analysis with Histograms

**Technology Sector:**

Market Capitalization (`market_cap`):

 - Average Market Cap: **R$ 8.72 billion**
 - Median Market Cap: **R$ 3.50 billion**
   - Indicates the substantial influence of large firms skewing the average, while most companies maintain a more modest capitalization.

Total Revenue (`total_revenue`):

 - Average Total Revenue: **R$ 3.43 billion**
 - Median Total Revenue: **R$ 1.50 billion**
   - Shows that while some companies lead with significantly high revenues, the distribution is balanced among most sector participants.

**Financial Sector:**

Profit Margin (`profit_margin`):

 - Average Profit Margin: **24.67%**
 - Median Profit Margin: **19.50%**
   - Reflects the trend of companies in this sector to maintain consistent profitability.

Operating Margin (`operating_margin`):

 - Average Operating Margin: **28.78%**
 - Median Operating Margin: **23.45%**
   - Highlights the sector's ability to maintain good margins despite market fluctuations.

**Industrial Sector:**

Market Capitalization (`market_cap`):

 - Average Market Cap: **R$ 58.50 billion**
 - Median Market Cap: **R$ 17.10 billion**
   - Indicates a broad range of company sizes within the sector, from smaller firms to industry giants.

Total Revenue (`total_revenue`):

 - Average Total Revenue: **R$ 9.33 billion**
 - Median Total Revenue: **R$ 4.30 billion**
   - Suggests a solid and consistent financial performance throughout the sector.

**Consumer Sector:**

Profit Margin (`profit_margin`):

 - Average Profit Margin: **12.45%**
 - Median Profit Margin: **8.50%**
   - Shows volatility and variation in pricing and cost strategies.

Operating Margin (`operating_margin`):

 - Average Operating Margin: **15.70%**
 - Median Operating Margin: **10.75%**
   - Indicates varied operational practices within the sector, with some firms achieving notable efficiencies.

**Conclusion:**

Through detailed histogram analysis and numerical value assessment by sector, substantial variations reflecting the competitive and operational dynamics of the analyzed companies have been identified. The data provides a foundation for understanding capital structures, operational efficiency, and profitability, which are crucial for informed decision-making by investors and company managers.

#### Market Cap

In [35]:
sector_means = df_fundamentals_numeric_cols.groupby(df_fundamentals_book['sector']).mean()
sector_means = sector_means.sort_values(by=['market_cap', 'total_revenue'], ascending=[True, False])

fig = px.bar(sector_means, x=sector_means.index, y='market_cap', title='Average Market Cap by Sector', color_discrete_sequence=['rgb(100, 195, 181)'], template='plotly_dark', hover_name=sector_means.index)
formatted_market_cap = [f'{x:.2f}B' for x in (sector_means['market_cap'] / 1e9)]
fig.update_traces(text=formatted_market_cap, textposition='outside',textfont=dict(color='white'))
fig.update_layout(xaxis_title='Setor', yaxis_title='Average Market Cap', height = 520)
fig.show()

- **`Consumer Cyclical`:** The sector with the largest number of listed companies, totaling 53, has the lowest average market cap at R$ 2.03 billion.
- **`Real Estate`:** Has an average market cap slightly higher at R$ 2.18 billion, with equal representation as the Consumer Defensive sector, each having 21 listed companies.
- **`Technology`:** Although being the least represented with only four listed companies, it has a significant average market cap of R$ 7.08 billion.
- **`Healthcare`:** With 13 companies, the sector shows an average market cap of R$ 7.21 billion.
- **`Industrials`:** Near the top in terms of company count with 48 listed, holds an average market cap of R$ 11.40 billion.
- **`Communication Services`:** A bit more represented than Technology with seven companies, shares a similar average market cap with Industrials at R$ 11.43 billion.
- **`Consumer Defensive`:** Along with Real Estate, both sectors boast identical representation and a substantial average market cap of R$ 14.78 billion.
- **`Utilities`:** Surpassing Financial Services in company count with 46 listed, the sector's average market cap stands at R$ 15.78 billion.
- **`Basic Materials`:** With 29 companies, this sector has an average market cap of R$ 20.82 billion.
- **`Financial Services`:** Having a strong representation with 40 companies, it reaches an average market cap of R$ 39.92 billion.
- **`Energy`:** With only nine companies listed, it is distinguished by the highest average market cap at R$ 118.12 billion.

The graphic analysis suggests that the Consumer Cyclical sector, despite having the highest company count, might reflect diversity and strength but faces a challenging market valuation context. Meanwhile, the Technology sector’s lower representation may indicate less market maturity or growth potential in the Brazilian stock exchange. Energy, with the highest average market cap, indicates substantial sector size and investor confidence.


#### Profitability by Sector

In [36]:
sector_means = sector_means.sort_values(by=['operating_margins'], ascending=True)

fig_op_margin = px.bar(x=sector_means.index, y=sector_means['operating_margins'], color_discrete_sequence=['rgb(100, 195, 181)'], hover_name=sector_means.index)
fig_op_margin.update_traces(text=sector_means.operating_margins.round(2)*100, textposition='outside',textfont=dict(color='white'))
fig_op_margin.update_layout(title='Average Operating Margin by Sector', xaxis_title='Sector', yaxis_title='Operating Margin (%)', template = 'plotly_dark')
fig_op_margin.show()

- **`Real Estate`:** Operational margin deeply negative at **-61%**, signifying considerable challenges in profit generation after covering operational costs.
- **`Consumer Cyclical`:** Similarly distressing with a very negative operational margin at **-60%**, indicating potential struggles with high operating costs relative to revenue.
- **`Industrials`:** A negative operational margin at **-26%** signals average industrial companies are experiencing operational difficulties leading to pre-interest and tax losses.
- **`Energy`:** Marginally negative operational margin at **-3%**, suggesting the sector is near breakeven but still has operational costs outpacing revenue.
- **`Communication Services`:** Low but positive operational margin at **1%**, reflecting a modest average operational profit after costs.
- **`Healthcare`:** A healthy operational margin at **8%**, indicating profitable operations within the sector.
- **`Consumer Defensive`:** With an operational margin of **10%**, this sector is generating healthy operational profits over revenue.
- **`Basic Materials`:** Matching the Consumer Defensive with a **10%** margin, indicating strong operational performance in material processing and production.
- **`Technology`:** High operational margin at **16%**, demonstrating the sector's ability to maintain efficient and profitable operations.
- **`Utilities`:** With a significant high operational margin at **26%**, this suggests robust profit generation from regular operations and lesser vulnerability to business environment fluctuations.
- **`Financial Services`:** The sector boasts the highest operational margin at **28%**, signifying an excellent operational performance and efficiency in profit generation pre-financing and tax.

The analysis indicates a wide variance in operational efficiency across sectors, with Real Estate and Consumer Cyclical facing the most significant challenges, while Financial Services and Utilities exhibit strong operational profitability. Technology also stands out for its high efficiency, despite its smaller size in terms of market representation.


In [37]:
sector_means = sector_means.sort_values(by=['profit_margins'], ascending=True)

fig_net_margin = go.Figure()
fig_net_margin.add_trace(go.Bar(x=sector_means.index, y=sector_means['profit_margins'], name='Net Margin', marker_color='rgb(100, 195, 181)'))
fig_net_margin.update_traces(text=[f'{x:.2f}%' for x in (sector_means['profit_margins'])], textposition='outside',textfont=dict(color='white'))
fig_net_margin.update_layout(height=555,title='Average Net Margin by Sector', xaxis_title='Setor', yaxis_title='Net Margin(%)', template='plotly_dark')

fig_net_margin.show()

- **`Communication Services`:** Extremely negative net margin at -0.57%, implying average losses after all expenses.
- **`Healthcare`:** Break-even net margin at 0.01%, indicating the sector is just managing to cover its costs.
- **`Consumer Defensive`:** Positive yet low net margin at 0.04%, showing some profitability.
- **`Basic Materials`:** Also a low net margin at 0.08%, meaning that despite operational profits, the final profitability remains modest.
- **`Consumer Cyclical`:** Comparable net margin to Basic Materials at 0.09%, reflecting slight profitability after all expenses.
- **`Technology`:** With a net margin of 0.16%, it's better than the preceding sectors but still modest.
- **`Energy`:** Net margin of 0.21%, showing that energy companies can maintain reasonable profitability after all costs.
- **`Financial Services`:** A higher net margin at 0.34%, demonstrating that this sector is more profitable compared to many others.
- **`Utilities`:** Solid net margin at 0.34%, on par with Financial Services, indicative of strong profitability.
- **`Industrials`:** High net margin at 0.64%, one of the highest, suggesting that industrial companies are quite profitable after all expenses.
- **`Real Estate`:** The highest net margin reported at 1.20%, marking it as the most profitable sector among those listed after accounting for all expenses.

This overview reveals a significant spectrum in terms of profitability with the Real Estate sector leading in terms of net margin, while Communication Services struggle with net losses. The analysis suggests a robust profitability for sectors like Real Estate and Industrials post all deductions, whereas sectors like Healthcare operate at a no-gain, no-loss basis.


In [38]:
sector_means = sector_means.sort_values(by=['roce'], ascending=True)

fig_roce = go.Figure()
fig_roce.add_trace(go.Bar(x=sector_means.index, y=sector_means['roce'], name='ROCE', marker_color='rgb(100, 195, 181)',text=sector_means['roce'].apply(lambda x: f'{x:.2f}%'), textposition='outside', textfont=dict(color='white')))
fig_roce.update_layout(height=550, title='Return on Capital Employed (ROCE) by Sector', xaxis_title='Sector', yaxis_title='ROCE (%)', template='plotly_dark')
fig_roce.show()

**Negative Returns**
- **`Energy`**: With a ROCE of -0.62%, the sector indicates a loss on the capital employed on average. Factors could be volatile commodity prices, large capital investments not yet yielding returns, or inefficient operations.

- **`Real Estate`**: A ROCE of -0.06% suggests the real estate sector is also earning a negative return on capital, possibly due to a down market, high financing costs, or slower property appreciation rates.

**Break-even Performance**
- **`Consumer Cyclical`***: At a ROCE of 0.00%, companies in this sector are breaking even in terms of capital employed profitability, reflecting the sector's sensitivity to economic cycles.

**Modest Positive Returns**
- **`Communication Services`**: With a positive ROCE of 0.07%, this sector is generating a small but positive return on invested capital, covering a diverse range of companies in telecommunications and media.

- **`Healthcare`**: Exhibiting a ROCE of 0.08%, the healthcare sector achieves a slightly positive return, indicating demand stability for health services and products despite high entry barriers and significant technology and research investments.

**Reasonable Efficiency**
- **`Industrials`**: A ROCE of 0.10% suggests reasonable capital use efficiency within a diverse sector encompassing construction, machinery, transportation, and more.

**Above Average Performance**
- **`Consumer Defensive`** and **`Technology`**: Both sectors show a ROCE of 0.12%. Consumer Defensive typically includes less economically sensitive companies like food and beverages, while Technology balances rapid innovation and growth against significant risks and capital investments.

- **`Utilities`**: A ROCE of 0.14% indicates above-average performance for utilities, which often operate under regulatory frameworks ensuring certain return levels.

**Leading Sectors**

- **`Financial Services`** and **`Basic Materials`**: Leading with a ROCE of 0.15%, Financial Services include high-leverage and capital-efficient businesses like banks and insurance, while Basic Materials may benefit from favorable market prices or efficient operations.


The chart shows that ROCE, an indicator of the efficiency with which capital is employed, varies across sectors. The energy sector has the lowest return, indicating challenges in generating profit from the capital used. In contrast, the basic materials, financial services, and utilities sectors lead with the highest ROCE, showing more effective capital management. The other sectors fall between these extremes, with returns ranging from slightly negative to moderately positive.

#### Debt Analysis by Sector

In [39]:
df = pd.DataFrame()
df['total_debt'] = pd.to_numeric(df_fundamentals_book['total_debt'], errors='coerce')
df['ebitda'] = pd.to_numeric(df_fundamentals_book['ebitda'], errors='coerce')
df['total_equity'] = pd.to_numeric(df_fundamentals_book['book_value'], errors='coerce')

sector_debt_ebitda = df.groupby(df_fundamentals_book['sector'])['total_debt'].sum() / df.groupby(df_fundamentals_book['sector'])['ebitda'].sum()
sector_debt_ebitda.sort_values(ascending=True, inplace=True)
sector_debt_equity = df.groupby(df_fundamentals_book['sector'])['total_debt'].sum() / df.groupby(df_fundamentals_book['sector'])['total_equity'].sum()
sector_debt_equity.sort_values(ascending=True, inplace=True)

fig = sp.make_subplots(rows=2, cols=1, subplot_titles=('Debt/EBITDA by Sector', 'Debt/Equity by Sector'))

fig.add_trace(go.Bar(x=sector_debt_ebitda.index, y=sector_debt_ebitda, name='Debt/EBITDA Ratio', marker_color='rgb(100, 195, 181)', text=['R$ {:,.2f}'.format(val) for val in sector_debt_ebitda], textposition='outside', textfont=dict(color='white')), row=1, col=1)
fig.add_trace(go.Bar(x=sector_debt_equity.index, y=sector_debt_equity, name='Debt/Equity Ratio', marker_color='rgb(100, 195, 181)', text=['R$ {:,.2f}'.format(val) for val in sector_debt_equity], textposition='outside', textfont=dict(color='white')), row=2, col=1)

fig.update_layout(title_text='Sector Debt Analysis', height=1100, showlegend=False, template='plotly_dark')
fig.show()


**Debt/EBITDA by Sector**

- **`Energy`**: With a debt/EBITDA ratio of **1.09**, suggests being the least leveraged sector, indicating a lower dependence on debt relative to operational earnings.
- **`Basic Materials`**: A debt/EBITDA ratio of **2.09**, which shows a moderate use of financial leverage in relation to the operational cash generation capacity.
- **`Technology`**: Presents a ratio of **2.28**, which may indicate a healthy balance between the use of debt and the generation of EBITDA.
- **`Utilities`**: With a ratio of **3.50**, this sector may be utilizing a moderate to high level of debt in its capital structure.
- **`Healthcare`**: Records a ratio of **4.34**, possibly reflecting a greater leverage to finance operations and investments in the sector.
- **`Consumer Defensive`**: The ratio of **4.51** suggests a considerable use of debt compared with the EBITDA generated by the sector.
- **`Industrials`**: With a ratio of **5.07**, indicates that the sector may be relying on a relatively high level of debt for its operations.
- **`Communication Services`**: A ratio of **5.27** could demonstrate a significant dependence on debt financing relative to operational profit generation.
- **`Real Estate`**: Has a debt/EBITDA ratio of **5.89**, which is common for the sector given the high investment cost in properties.
- **`Consumer Cyclical`**: With a ratio of **6.07**, may suggest a growth strategy supported by higher levels of debt.
- **`Financial Services`**: Shows an anomalously high ratio of **224.70**, highlighting an extremely leveraged capital structure that is typical for this sector, where debt is a central instrument for banking and financial operations.


The chart presents the debt-to-EBITDA ratio by sector, where sectors such as energy have low leverage (1.09) and financial services have extremely high leverage (224.70), indicating contrasting approaches in debt management and operations.

**Debt/Equity by Sector**

- **`Energy`**: Possesses a negative value of R$ -348,021,372.87, which may suggest the sector has more equity than debt or that significant amortization has been applied.
- **`Consumer Cyclical`**: Shows a debt-to-equity balance of R$ -19,074,066.41, indicating more equity than debt.
- **`Real Estate`**: Displays R$ 13,038,292.98, pointing to a moderate level of debt in relation to equity.
- **`Healthcare`**: With a figure of R$ 350,146,740.16, demonstrates a substantial use of debt compared to equity.
- **`Technology`**: Carries a balance of R$ 377,696,184.76, indicating a considerable reliance on debt.
- **`Utilities`**: Debt over equity stands at R$ 720,574,425.15, showing significant financial leverage.
- **`Basic Materials`**: Has a level of indebtedness of R$ 735,778,874.33 over equity.
- **`Consumer Defensive`**: Presents R$ 962,523,154.85, signaling high indebtedness compared to equity.
- **`Industrials`**: With R$ 2,453,404,192.52, suggests a heavy use of debt.
- **`Financial Services`**: With R$ 5,240,302,297.02, is one of the sectors with a high debt-to-equity ratio.
- **`Communication Services`**: Has the highest balance of R$ 149,421,020,783.51, indicating very high leverage against equity.

These numbers show the diverse capital structure across sectors, with some sectors operating with much higher debt levels relative to their equity.


#### Dividend Distribution by Sector

In [40]:
sector_dividend_rate_avg = df_fundamentals_book.groupby('sector')['dividend_rate'].mean()
sector_dividend_rate_avg.sort_values(ascending=True, inplace=True)
sector_dividend_yield_avg = df_fundamentals_book.groupby('sector')['trailing_annual_dividend_yield'].mean()
sector_dividend_yield_avg.sort_values(ascending=True, inplace=True)

fig = sp.make_subplots(rows=2, cols=1, subplot_titles=('Average Dividend Yield by Sector', 'Average Dividend Yield by Sector'))

fig.add_trace(go.Bar(x=sector_dividend_rate_avg.index, y=sector_dividend_rate_avg, name='Dividend Rate', marker_color='rgb(100, 195, 181)', text=sector_dividend_rate_avg.apply(lambda x: f'R${x:,.2f}'), textposition='outside', textfont=dict(color='white')), row=1, col=1)
fig.add_trace(go.Bar(x=sector_dividend_yield_avg.index, y=sector_dividend_yield_avg, name='Dividend Yield', marker_color='rgb(100, 195, 181)', text=sector_dividend_yield_avg.apply(lambda x: f'{x:.2f}%'), textposition='outside', textfont=dict(color='white')),row=2, col=1)

fig.update_layout(title_text='Dividend Distribution by Sector', height=1000, showlegend=False, template='plotly_dark')
fig.show()

**Average Dividend Yield by Sector**

- **`Healthcare` (R$0.22)**: Lowest average dividend yield, reflecting possible reinvestment of profits into research and development.
- **`Technology` (R$0.33)**: Modest yields, possibly due to a focus on growth and innovation within the sector.
- **`Communication Services` (R$0.41)**: Slightly higher dividends, indicating stability and consistent income flows in the sector.
- **`Non-Cyclical Consumer` (Defensive) (R$0.80)**: Safer yield, typical of sectors with stable demand regardless of economic conditions.
- **`Industrials` (R$0.85)**: Dividends represent a balance between reinvestment and return to shareholders in a diversified sector.
- **`Financial Services` (R$1.05)**: Indicates a traditional profit distribution, aligned with the return expectations of a regulated and stable sector.
- **`Cyclical Consumer ` (R$1.71)**: Attractive dividends, reflecting the ability to generate profits in times of economic upturn.
- **`Energy` (R$1.97)**: Robust yields, suggesting a sector with solid cash flows and generous distribution policies.
- **`Utilities` (Public Services) (R$2.41)**: High dividends, typical of sectors with guaranteed demand and predictable income streams.
- **`Basic Materials` (R$3.44)**: High dividends may reflect commodity price volatility and a desire to attract investors.
- **`Real Estate` (R$381.78)**: Extraordinarily high dividend yield, highlighting the real estate sector as likely including REITs, which are mandated to distribute most of their profits to shareholders.

Based on this data, we can see that the Real Estate sector has an exceptionally high average dividend yield compared to other sectors. This could be the result of a number of factors such as aggressive profit distribution policies, significant capital yields, or a discrepancy between the share price and the dividend paid. However, this value is much higher than those of other sectors and could indicate an atypical situation or an error in the chart since it is unusually higher than the yields of other sectors by a substantial margin.

**Average Dividend Yield by Sector**

- **`Healthcare` (0.01%)**: This sector shows the lowest dividend yield, suggesting that companies might be reinvesting earnings rather than distributing them to shareholders.
- **`Real Estate` (0.02%)**: Despite the significant value previously shown in reais, the percentage is low, which might indicate a large variation in dividend amounts or share prices.
- **`Cyclical Consumer` (0.02%)**: The dividend yields are low, which might reflect a reinvestment policy or profit variation with economic swings.
- **`Industrials` (0.02%)**: This sector shows moderate dividend yields, perhaps due to a mix of reinvestment and dividend payments across varied industries.
- **`Non-Cyclical Consumer` (Defensive) (0.03%)**: The sector offers a slightly higher yield, possibly due to its stability and less sensitivity to economic fluctuations.
- **`Communication Services` (0.03%)**: The yield is on par with defensive consumer goods, which might indicate consistent cash flows in the sector.
- **`Financial Services `(0.03%)**: The yield reflects the norm for a sector that balances dividend payments with earnings retention for financial stability.
- **`Technology` (0.05%)**: A higher yield in this chart, suggesting the sector may be maturing, with some companies beginning to distribute more profits as dividends.
- **`Utilities` (0.06%)**: Traditionally, this sector is known for higher dividend payments due to the predictable and regulated nature of its cash flows.
- **`Energy` (0.06%)**: This sector presents a robust dividend yield, which might reflect established companies with profitable operations.
- **`Basic Materials` (0.08%)**: The sector with the highest dividend yield on the chart, indicating that companies in this sector may be generating substantial profits and opting to distribute a significant portion to shareholders.

This percentage distribution of dividends provides an alternative view to the one presented earlier, highlighting how dividend yields can vary widely between sectors and the importance of considering both the absolute value and the yield relative to the stock price.


#### Valuation by Sector

In [41]:
sector_pe_avg = df_fundamentals_book.groupby('sector')['trailing_pe'].mean().sort_values()
sector_pb_avg = df_fundamentals_book.groupby('sector')['price_to_book'].mean().sort_values()

fig = sp.make_subplots(rows=2, cols=1, subplot_titles=('Average P/E Ratio by Sector', 'Average P/B Ratio by Sector'))
fig.add_trace(go.Bar(x=sector_pe_avg.index, y=sector_pe_avg, name='P/E Ratio', marker_color='rgb(100, 195, 181)', text=sector_pe_avg.apply(lambda x: f'{x:.2f}x'), textposition='outside', textfont=dict(color='white')), row=1, col=1)
fig.add_trace(go.Bar(x=sector_pb_avg.index, y=sector_pb_avg, name='P/B Ratio', marker_color='rgb(100, 195, 181)', text=sector_pb_avg.apply(lambda x: f'{x:.2f}x'), textposition='outside', textfont=dict(color='white')), row=2, col=1)
fig.update_layout(title_text='Sector Valuation: P/E and P/B Ratios', height=1100, showlegend=False, template='plotly_dark')
fig.show()

**Average P/E Ratio by Sector**

- **`Energy` (5.21x):** This sector has the lowest P/E ratio, suggesting it may be undervalued or that the market expects slower earnings growth in this sector.
- **`Consumer Cyclical` (7.93x):** A relatively low P/E ratio may indicate that stocks are traded at lower prices relative to earnings, possibly due to uncertain growth expectations given their dependency on economic conditions.
- **`Non-Cyclical Consumer` (8.85x):** As this sector is less sensitive to economic swings, the moderate P/E suggests a balanced valuation between growth and risk.
- **`Real Estate` (9.32x):** The P/E in this sector may reflect a mix of steady growth and real estate speculation, depending on interest rates and the economy.
- **`Industrials` (10.54x):** A mid-range P/E suggests the market may expect steady growth, neither overly optimistic nor pessimistic.
- **`Communication Services` (10.68x):** This sector, including media and telecom companies, where the P/E indicates moderate earnings growth expectations.
- **`Financial Services` (10.72x):** A P/E similar to communication services, reflecting moderate growth expectations in a sector influenced by interest rates and regulations.
- **`Utilities` (12.65x):** A higher P/E indicates that stocks may be pricier relative to current earnings, perhaps due to the perceived stability of cash flows in this sector.
- **`Healthcare`(15.86x):** This sector has a higher P/E, suggesting investors are willing to pay more for shares due to expectations of profit growth and innovation in the sector.
- **`Basic Materials` (20.01x):** The relatively high P/E indicates that the sector's stocks might be valued for expected growth or the volatility of commodity prices.
- **`Technology` (63.04x):** With the highest P/E, the technology sector exhibits the greatest expectations for future earnings growth, reflecting market optimism about ongoing innovation and expansion.

P/E ratios provide an overview of how different sectors are valued by the market in terms of earnings growth and risk.


**Average P/B Ratio by Sector**

- **`Communication Services` (0.70x)**: This sector has the lowest P/B ratio, indicating that the market values it at less than its book value, which could suggest undervaluation or a sector that is out of favor.
- **`Real Estate` (0.90x)**: The P/B ratio is below 1, which might indicate that the assets are valued less in the market than on the books, or there could be skepticism about the value of those assets.
- **`Financial Services` (1.10x)**: Close to 1, suggesting that the market valuation is in line with the book value of the assets.
- **`Consumer Cyclical` (1.25x)**: The market is willing to pay a modest premium over the book value, which might suggest expectations for growth that align with economic cycles.
- **`Energy` (1.28x)**: Slightly higher, indicating a reasonable market premium over book value, reflecting potential growth or recovery in the sector.
- **`Basic Materials` (1.50x)**: A higher P/B ratio could be due to the market pricing in the profitability of raw materials or future growth expectations.
- **`Consumer Defensive` (1.60x)**: This suggests that investors are willing to pay a bit more over the book value, likely due to stable earnings and dividends in this sector.
- **`Healthcare` (2.08x)**: A P/B ratio over 2 indicates a significant premium that investors are willing to pay, likely due to high expectations for future growth in the sector.
- **`Technology` (2.16x)**: Reflecting high growth expectations and possibly a lot of intangible assets that are highly valued, like intellectual property.
- **`Utilities` (8.65x)**: A very high P/B ratio, unusual for what is typically a stable sector; this might indicate that the book values are very low or the market expects significant growth or restructuring.
- **`Industrials` (19.56x)**: The highest P/B ratio, suggesting a very high premium that investors are willing to pay over the book value, which could indicate a strong future outlook or a sector that has many assets undervalued on its books compared to the market value.

These P/B ratios provide a snapshot of how the market values the assets of companies in different sectors, with higher ratios indicating more expensive valuations relative to book value.

#### Efficiency by Sector

In [42]:
df['return_on_assets'] = pd.to_numeric(df_fundamentals_book['return_on_assets'], errors='coerce')
df['total_revenue'] = pd.to_numeric(df_fundamentals_book['total_revenue'], errors='coerce')
df['total_assets'] = pd.to_numeric(df_fundamentals_book['total_assets_approx'], errors='coerce')
df['asset_turnover'] = df_fundamentals_book['total_revenue'] / df_fundamentals_book['total_assets_approx']

sector_efficiency = df_fundamentals_book.groupby('sector').agg({'return_on_assets': 'mean', 'asset_turnover': 'mean'}).reset_index()
sector_efficiency.sort_values(by='return_on_assets', ascending=True, inplace=True)

fig1 = go.Figure()
fig1.add_trace(go.Bar(x=sector_efficiency['sector'], y=sector_efficiency['return_on_assets'], name='ROA', marker_color='lightseagreen', text=sector_efficiency['return_on_assets'].apply(lambda x: f'{x:.2f}%'), textposition='outside'))
fig1.update_layout(title='Efficiency by Sector: Return on Assets (ROA)',xaxis_tickangle=-45, barmode='group', template='plotly_dark', height=600)
fig1.show()

sector_efficiency.sort_values('asset_turnover', ascending=True, inplace=True)

fig2 = go.Figure()
fig2.add_trace(go.Bar(x=sector_efficiency['sector'], y=sector_efficiency['asset_turnover'], name='Giro dos Ativos', marker_color='lightseagreen', text=sector_efficiency['asset_turnover'].apply(lambda x: f'{x:.2f}'), textposition='outside'))
fig2.update_layout(title='Efficiency by Sector: Asset Turnover Ratio', xaxis_tickangle=-45, barmode='group', template='plotly_dark', height=600)
fig2.show()

**Efficiency by Sector: Return on Assets (ROA)**

- **`Communication Services` (-0.00%)**: This sector is showing a 0% ROA, which suggests no return on assets or a very marginal profit that rounds down to zero, indicating that the sector, on average, is not generating a profit from its assets.
- **`Real Estate` (0.02%)**: A very low ROA, indicating minimal profitability in relation to assets. This could be due to high asset base and low income or inefficient asset use.
- **`Financial Services` (0.03%)**: Similarly low, which could be surprising given that this sector's business revolves around asset and capital management. However, it could also reflect large asset bases typical for this sector.
- **`Consumer Cyclical` (0.03%)**: This is typically a sector with varying performance that is highly dependent on the economic cycle. A low ROA here indicates low profitability relative to the assets employed.
- **`Technology` (0.04%)**: Moderate ROA, which might be due to the high asset turnover or efficient asset usage, but not particularly high in profitability.
- **`Consumer Defensive` (0.04%)**: The same as technology, this sector has a moderate ROA, possibly due to stable earnings but relatively high assets.
- **`Healthcare` (0.04%)**: Consistent with the technology and consumer defensive sectors, indicating moderate efficiency in asset use.
- **`Energy` (0.04%)**: Similar to healthcare, it's a moderate figure, potentially indicating substantial investment in assets is required for the returns generated.
- **`Basic Materials` (0.05%)**: A little higher, suggesting better profitability per asset or more efficient use of assets.
- **`Industrials` (0.06%)**: Higher ROA could indicate better asset utilization or a more profitable sector on average compared to others.
- **`Utilities `(0.07%)**: The highest ROA on this chart, suggesting this sector is most efficient at generating profit from its assets, which might reflect regulated returns on assets or effective management.

The sector ROA figures could inform investment decisions and provide insights into the efficiency of asset use in each sector. It's important to consider these figures in the context of each industry's unique characteristics and economic environment.

**Efficiency by Sector: Asset Turnover Ratio**

There seems to be an anomaly with the 'Energy' sector showing an extremely high value (98525.46) which is several orders of magnitude larger than the other sectors. This is likely an error or an outlier caused by data misrepresentation or an extraordinary event within the sector that isn't typical for the performance of that sector, such as the sale of a significant asset or assets which generated revenue not typical for the operational business.

- **`Financial Services` (2.61)**: Indicates a heavy asset base relative to revenues, typical for the sector due to holding significant financial assets.
- **`Communication Services` (4.26)**: Shows a more efficient use of assets to generate revenue than financial services.
- **`Real Estate` (5.82)**: Reflects the revenue-generating capacity of real estate investments.
- **`Technology `(6.15)**: Suggests efficient use of assets for generating sales revenue.
- **`Utilities `(6.46)**: Demonstrates a balance between large infrastructure investments and revenue.
- **`Healthcare `(16.42)**: Indicates efficient use of assets in revenue generation.
- **`Industrials` (22.73)**: Signifies high efficiency in asset turnover, which may point to lower asset intensity or higher sales volume.
- **`Consumer Defensive `(27.70)**: Shows very efficient asset use, likely due to less capital-intensive operations.
- **`Basic Materials` (107.97) and Consumer Cyclical (120.19)**: These sectors exhibit extremely efficient asset turnover, suggesting either lower reliance on heavy assets or higher sales volume relative to asset values.
- **`Energy` (98525.46)**: An anomalously high ratio, indicating a potential outlier or error that necessitates further investigation.

When analyzing asset turnover ratios, it is crucial to take into account sector-specific operational and business model differences. Comparisons should be made against historical sector performance or within the same sector to provide context and understanding.


#### Risk Analysis by Sector

In [43]:
sector_volatility = df_fundamentals_book.groupby('sector')['return_on_assets'].std().reset_index()
sector_volatility.sort_values(by='return_on_assets', ascending=True, inplace=True)

fig1 = go.Figure()
fig1.add_trace(go.Bar(x=sector_volatility['sector'], y=sector_volatility['return_on_assets'], name='Volatility', marker_color='lightseagreen', text=sector_volatility['return_on_assets'].apply(lambda x: f'{x:.2f}'), textposition='outside'))
fig1.update_layout(title='Volatility (Standard Deviation) of ROA by Sector', xaxis_tickangle=-45, template='plotly_dark', height=550)
fig1.show()


**Volatility (Standard Deviation) of ROA by Sector**

- **`Technology` (0.03)**: Exhibits the lowest volatility, suggesting stable sector ROA.
- **`Communication Services`(0.03)**: Tied with technology for low volatility, indicating stability in asset returns.
- **`Real Estate` (0.04)**: Shows slightly higher volatility, consistent with the long-term and predictable nature of real estate returns.
- **`Consumer Defensive` (0.05)**: Often features companies with operations less affected by economic downturns, explaining the low to moderate volatility.
- **`Financial Services `(0.05)**: Moderate volatility, reflecting the impact of financial market fluctuations and interest rate changes on asset returns.
- **`Utilities` (0.06)**: Given the sector's stable customer base and regulated returns, it has moderate volatility.
- **`Industrials `(0.06)**: The volatility level is moderate due to the cyclical nature of the sector and variability in demand.
- **`Consumer Cyclical` (0.07)**: A higher level of volatility, correlating with economic cycle sensitivity.
- **`Healthcare` (0.08)**: Shows moderate to high volatility, affected by demand fluctuations, regulatory shifts, and innovation cycles.
- **`Basic Materials` (0.08)**: This sector's volatility can be attributed to the variable commodity prices and market conditions.
- **`Energy` (0.10)**: The highest volatility, reflecting sensitivity to commodity pricing, geopolitical tensions, and regulatory shifts.

Investors consider the volatility of ROA in their assessment of investment stability and risk. Higher volatility in ROA suggests a higher risk profile for the sector, which may warrant a greater expected return to compensate for the increased risk.

In [44]:
sector_debt_equity = df_fundamentals_book.groupby('sector')['debt_to_equity'].mean().reset_index()
sector_debt_equity.sort_values(by='debt_to_equity', ascending=True, inplace=True)

fig = go.Figure()
fig.add_trace(go.Bar(x=sector_debt_equity['sector'], y=sector_debt_equity['debt_to_equity'], name='Debt/Equity Ratio', marker_color='lightseagreen', text=sector_debt_equity['debt_to_equity'].apply(lambda x: f'{x:.2f}'), textposition='outside'))
fig.update_layout(height=550,title='Debt-to-Equity Ratio by Sector', xaxis_tickangle=-45, template='plotly_dark')
fig.show()


**Debt-to-Equity Ratio by Sector**

- **`Energy` (-5.89)**: Suggests more equity than debt or the influence of accounting adjustments or special financial structures.
- **`Real Estate` (-2.13)**: Indicates a propensity for equity financing or specific financial factors resulting in a negative D/E ratio.
- **`Utilities` (-1.16)**: Unusual for a sector known for heavy infrastructure financed typically through debt.
- **`Consumer Defensive` (-1.07)**: Points towards more equity financing or other exceptional factors affecting the D/E ratio.
- **`Healthcare `(-0.94)**: A negative D/E ratio which is peculiar for the sector.
- **`Technology` (-0.68)**: May reflect a strong balance sheet with a lesser dependence on debt financing.
- **`Industrials` (-0.22)**: The negative ratio is unusual and might suggest sector-specific financial structures.
- **`Consumer Cyclical` (0.59)**: Shows a balanced use of debt relative to equity, suggesting a healthy leverage level.
- **`Basic Materials` (3.33)**: A higher reliance on debt, indicative of the capital requirements to fund operations and growth.
- **`Communication Services` (6.46)**: Reflects significant debt financing, potentially due to the heavy capital expenditures in the sector.
- **`Financial Services` (28.73)**: Extremely high D/E ratio, typical due to the nature of financial operations involving higher debt levels.

Negative D/E ratios are atypical and warrant further investigation to understand the underlying causes, which may include negative shareholder equity or particular accounting practices. Investors should approach these figures with caution and seek to understand the specifics of each sector.


#### Raking dos setore

In [45]:
df_fundamentals_book['investment_score'] = df_fundamentals_book[['profit_margins', 'operating_margins', 'return_on_equity']].mean(axis=1)
investment_score_mean = df_fundamentals_book.groupby('sector')['investment_score'].mean().sort_values(ascending=True).reset_index()

fig = px.bar(investment_score_mean, x='sector', y='investment_score', title='Investment Score by Sector', labels={'investment_score': 'Investment Score (%)', 'sector': 'Setor'}, template='plotly_dark', height=550)
fig.update_traces(texttemplate='%{y:.2%}', textposition='outside', marker_color='lightseagreen')
fig.show()

**Investment Score by Sector**`

The evaluation of sectors varies significantly, with some sectors displaying negative investment scores, while others appear more attractive:

- **`Communication Services`**: -17.79%
  The lowest on the chart, indicating a negative outlook for investments.
- **`Consumer Cyclical`**: -16.21%
  Also negative, suggesting less favorable evaluations for investments.
- **`Healthcare`**: 2.15%
  A low positive score, indicating a slightly favorable investment outlook.
- **`Consumer Defensive`**: 3.45%
  Slightly better, indicating a modestly favorable view for investment.
- **`Basic Materials`**: 7.32%
  Mid-range positive score, suggesting some attractiveness for investment.
- **`Energy`**: 10.73%
  Above the midpoint in terms of investment attractiveness.
- **`Technology`**: 12.33%
  A higher score, reflecting a more favorable investment perspective.
- **`Industrials`**: 16.43%
  Among the more attractive sectors based on this score.
- **`Financial Services`**: 20.62%
  Even higher, indicating significant investment appeal.
- **`Real Estate`**: 25.59%
  One of the highest scores, suggesting a strong attractiveness for investment.
- **`Utilities`**: 33.03%
  The highest score, indicating it as the most attractive sector according to this criterion.


## TL/DR

Analyzing the financial data from various business sectors, we have two main perspectives: indebtedness and investment attractiveness.

**Debt levels**, measured by the debt-to-equity ratio, vary significantly across sectors. The financial sector exhibits the highest leverage, while energy, real estate, and utilities sectors show negative values, suggesting lower dependence on debt or other accounting specifics.

In terms of **investment attractiveness**, the utilities sector emerges as the most attractive, followed by real estate and financial services. In contrast, communication services and consumer cyclicals have the lowest scores, which could imply a perception of risk or low potential return by investors.

These insights suggest diversity in financial health and investment expectations among sectors, with some displaying high leverage and others being seen as more attractive for investments, not always correlating directly with their levels of indebtedness.
